File: backend/app/modules/issues/issues_service.py
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.future import select
from typing import List, Optional

from app.models.events.issue import Issues
from app.modules.issues.types.issues_dtos import PostIssueRequestDTO, IssueResponseDTO


class IssuesService:
    def __init__(self, db: AsyncSession):
        self.db = db

    async def create_issue(
            self,
            issue_data: PostIssueRequestDTO,
            author_id: int
    ) -> IssueResponseDTO:
        issue = Issues(
            name=issue_data.name,
            description=issue_data.description,
            longitude=issue_data.longitude,
            latitude=issue_data.latitude,
            author_id=author_id
        )

        self.db.add(issue)
        await self.db.flush()
        await self.db.commit()
        await self.db.refresh(issue)

        return self._to_response_dto(issue)

    async def get_issues_by_author(
            self,
            author_id: int
    ) -> List[IssueResponseDTO]:
        query = select(Issues).where(
            Issues.author_id == author_id,
            Issues.deleted_at.is_(None)
        )
        result = await self.db.execute(query)
        issues = result.scalars().all()

        return [self._to_response_dto(issue) for issue in issues]

    async def get_issue_by_id(
            self,
            issue_id: int
    ) -> Optional[IssueResponseDTO]:
        query = select(Issues).where(
            Issues.id == issue_id,
            Issues.deleted_at.is_(None)
        )
        result = await self.db.execute(query)
        issue = result.scalars().first()

        if not issue:
            return None

        return self._to_response_dto(issue)

    def _to_response_dto(self, issue: Issues) -> IssueResponseDTO:
        return IssueResponseDTO(
            id=issue.id,
            name=issue.name,
            description=issue.description,
            longitude=float(issue.longitude),
            latitude=float(issue.latitude),
            is_completed=issue.is_completed,
            created_at=issue.created_at.isoformat(),
            image_url=issue.image_url or "",
            author_id=issue.author_id
        )

File: backend/internal/modules/issues/issues.service.go
package issues

import (
	"errors"
	"github.com/DAR-Innovations/city-zen/internal/data"
	"github.com/DAR-Innovations/city-zen/internal/modules/issues/types"
	"gorm.io/gorm"
)

type IssuesService struct {
	db *gorm.DB
}

func NewIssuesService(db *gorm.DB) *IssuesService {
	return &IssuesService{db: db}
}

// CreateIssue creates a new issue record in the database
func (s *IssuesService) CreateIssue(authorID uint, dto *types.PostIssueRequestDTO, imagePath string) (uint, error) {
	// Save the issue to the database
	issue := &data.Issue{
		AuthorID:    authorID,
		Name:        dto.Name,
		Description: dto.Description,
		Longitude:   dto.Longitude,
		Latitude:    dto.Latitude,
		ImageURL:    imagePath,
	}

	if err := s.db.Create(&issue).Error; err != nil {
		return 0, err
	}

	return issue.ID, nil
}

// GetIssuesByAuthor retrieves all issues posted by a specific author
func (s *IssuesService) GetIssuesByAuthor(authorID int) ([]types.IssueResponseDTO, error) {
	var issues []data.Issue
	if err := s.db.Where("author_id = ?", authorID).Find(&issues).Error; err != nil {
		return nil, err
	}

	// Convert issues to DTOs
	var response []types.IssueResponseDTO
	for _, issue := range issues {
		response = append(response, types.IssueResponseDTO{
			ID:          issue.ID,
			Name:        issue.Name,
			Description: issue.Description,
			IsCompleted: issue.IsCompleted,
			Longitude:   issue.Longitude,
			Latitude:    issue.Latitude,
			ImageUrl:    issue.ImageURL,
			AuthorID:    issue.AuthorID,
		})
	}
	return response, nil
}

// GetIssueByID retrieves a specific issue by its ID
func (s *IssuesService) GetIssueByID(issueID string) (*types.IssueResponseDTO, error) {
	var issue data.Issue
	if err := s.db.Where("id = ?", issueID).First(&issue).Error; err != nil {
		return nil, errors.New("issue not found")
	}

	// Return the issue as a DTO
	return &types.IssueResponseDTO{
		ID:          issue.ID,
		Name:        issue.Name,
		Description: issue.Description,
		IsCompleted: issue.IsCompleted,
		Longitude:   issue.Longitude,
		Latitude:    issue.Latitude,
		ImageUrl:    issue.ImageURL,
		AuthorID:    issue.AuthorID,
	}, nil
}
